coll/tuned: Change the allreduce default collective algorithm selection

jiaxiyan opened the pull request:
Include the new allreduce algorithm allgather_reduce in the default algorithm selections. The default algorithm selections were out of date and not performing well. After gathering data using the ompi-collectives-tuning package, new default algorithm decisions are selected for allreduce, which have significant speedup for small node number and message size.

jiaxiyan commented:
You can access the detailed data for all algorithms [here](https://github.com/open-mpi/ompi/files/13790281/detail.txt)

wenduwan commented:
Could you also run allreduce benchmarks such as OMB/IMB?

rhc54 commented:
You'll probably need to get a bunch of people to run the benchmarks across a range of systems - past updates have sometimes proven to be rather system specific.

wenduwan commented:
@rhc54 I was also curious about this. A while ago I poked the community, and from what I heard the majority of the users have bespoke decision files, instead of using this decision tree.

I will tag more people after the holidays.

rhc54 commented:
> the majority of the users have bespoke decision files, instead of using this decision tree

Correct - issue has always been around updating the defaults since that impacts everyone who doesn't have a custom decision file. Hard to find a default decision tree that works equally well for all systems - main problem was that the proposed updates in the past actually hurt performance for some significant range of users.

Not saying an update isn't warranted - just cautioning that adoption can take considerable effort.

wenduwan commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -72,25 +72,42 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
-                alg = 1;
-            }
-        } else if (communicator_size < 8) {
+        if (communicator_size == 2) {
```
This logic only applies to disjoint communicators. FWICT communication patterns on the same node are faster with different algorithms.

bosilca commented on ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:
```diff
@@ -217,6 +217,189 @@ ompi_coll_tuned_allreduce_intra_dec_fixed(const void *sbuf, void *rbuf, int coun
 }
 
+
+/*
```
Add a more descriptive comment please

bosilca reviewed (commented):
The PR will be good to go as soon as it gets a better description of the newly added `ompi_coll_tuned_allreduce_intra_disjoint_dec_fixed` function.

jiaxiyan committed:
coll/tuned: Change the allreduce default collective algorithm selection

Include the new allreduce algorithm allgather_reduce in the default
algorithm selections. The default algorithm selections were out of date
and not performing well. After gathering data using the ompi-collectives-tuning
package, new default algorithm decisions are selected for allreduce, which
have significant speedup for small node number and message size on inter node
communicators.

Signed-off-by: Jessie Yang <jiaxiyan@amazon.com>

hppritcha commented:
anyone still working on this PR @jiaxiyan
